### ISSUE TITLE
Toggle alarm

### ISSUE THREAD

**Issue by bohtho** [perbrage/node-red-contrib-sectoralarm#2]
Would it be possible to implement arming/disarming too like [this python package for interacting with Sector Alarm](https://github.com/mgejke/asyncsector) has done ? I just learnt about Node-RED yesterday (coming from Home Assistant) and am only a novice when it comes to javascript but would love to help where I can. Would love to almost only rely on Node-RED.

**Comment by perbrage** [comment]
I had a look at the python link you provided, and it seems easy enough to implement. I will need to add the capability to my nodeJs project first (https://github.com/perbrage/sectoralarm), and then I can add it to this node-red-contrib.

**Comment by bohtho** [comment]
Ah. I didn’t realise that was implemented in another general library but that makes sense. Very cool. 

The python library also seems to rely on the central api site of Sector instead of the regional sites too.

Holler if you need testing or me to find out about something more!

**Comment by perbrage** [comment]
Adding the arming/disarming support to this contrib poses 3 different options of implementation, as a Site can have many arming/disarming codes for different purposes. For example, master code, children codes or house keeping etc.

So the NodeRED site-note could either:

1) Have only one code in options. A limitation, but secure and fast to implement.
2) The input of the Site node could take a JSON payload with Action and Code, but that leaves the management of the code to the end-user which might pose a security risk.
3) Make a code/name mapping in options. This gives all codes a name to use instead, and then the payload would be an Action and a CodeName. This enables use of different codes, but also keeps the codes secure. It comes with additional implementation, and I have to research a bit on how to do it properly.

As of now, I am leaning towards option 1, but I wonder if option 1 will suffice for your purposes?

**Comment by bohtho** [comment]
All 3 options are an awesome improvement, seeing that you manged to add arming/disarming to the nodejs lib 👍 

Option 1 is the most user-friendly and an acceptable limitation (and would work for me). And I see implementations in other home automation software and languages also rely on one code.

I'm not sure I understand the security risk of option 2 (which is negated in option 3), but anyway it's more complicated to end-users in general (and to code) I guess.

**Comment by perbrage** [comment]
Option 1 implemented and pushed (v1.1.0). I also added new outputs (action and error), together with a history record limitation setting. Some bug fixing and polishes as well, so give it a try and let me know if you find any issues.

**Comment by bohtho** [comment]
I tested it just now. Looks polished, single, or repeatedly checking status works, arming works (with an informative "working..." message) great, but I have two issues (one very minor, and one bigger):

- ~~I made a new code at Sector especially for my Raspberry Pi (to be able to identify it), but when arming (partially) the debug payload still says the main users name. In the repeating status check however it says the correct users name.~~

- ~~Sending "disarm" (according to info) seems to do nothing. No error message, no "working..." status on node, no change in status. I've checked my Sector app to see if I have any strange settings for disallowing this but to no avail.~~

EDITED: because I made an idiotic node-red mistake. And concerning item 1 it seems it now reports the correct new code's name every time for every status. May have been a Sector or node-red issue.

So to justify bothering you with "faults" I have just one little thing to report: A spelling error in the status object : "partialArmingAvailabile" :) Sorry. Great work !

**Comment by perbrage** [comment]
Hi.

Many things to reply to here, so let's start. :)

a) Concerning the users name that arms/disarms this seems to be very odd on Sector Alarms side. I also noticed it is not always correct, and that at times it swaps over to just stating Code (or Kod in Swedish even). Sector Alarm API also only provides the user name in the history call, so when checking the actual status of the alarm, I have to fetch the history as well, and copy the user name from the most recent event. This could lead to an out of sync depending on their code, but I haven't seen it yet anyways. 

Arming/Disarming responses from Sector Alarm does not provide the user name, as it only states success / fail with some Alarm ID info.
For your info, I have hardcoded to translate from username Kod to Code in the general library. This might not be perfect, but it suits the Swedish users anyways.

b) I double checked the code after you wrote this about disarm, and I noticed that I had introduced a Copy&Paste AND a Search&Replace issue in v1.1.0. Basically, the issue would break session timeouts, where the relogin and the fulfilment of the command would not go through. To the end-user this would look like nothing is happening. But it should have produced a NodeRED node error (not the error output), but that is not visible unless you use a error catching node.
If that error would make the node invalid and not working at all, I don't know. But I would expect that subsequent calls would go through.
Also, arm would call partialArm during a session timeout retry.

c) partialArmingAvailabile originates from a spelling error of Sector Alarm API that transfered into the general library. Their property is even name partialAvalible, I guess it was too bad of a spelling error I couldn't even fix it when transforming the output, haha. I can solve this during the transformation in the general library. 

I am glad you are happy with the results. Be sure to update the package by the end of the day, as I will try to nail down the various issues mentioned above.

Thank you for the help!

**Comment by perbrage** [comment]
Make sure you are using v1.1.3 for any further testing :)

**Comment by bohtho** [comment]
I dont know if I have messed up my node-red or if this is a bug when having inserted the new sector alarm node and deploying: `"Credential type 'sectoralarm-site' is not registered"`

The alert window says: `Flows stopped due to missing node types. sectoralarm-site` even though the node is inserted.

**Comment by perbrage** [comment]
I would uninstall, make sure it is completly gone in your files in your .nodered folder, delete the credentials file and then install again.

**Comment by perbrage** [comment]
I see no issues using the latest version here as of now anyways, but can investigate later if you are still having issues

**Comment by perbrage** [comment]
Did you solve your last problem? About to close this issue if everything seems fine to you.

**Comment by bohtho** [comment]
I've just managed to fuddle up my node-red setup completely, so I couldn't say 100% the issue was gone before I've fixed it. But it sure seems everything I was wished for this node is implemented and working. Thank you very much! I think I'm just going to export my flows and bulldoze my setup and set-up node-red and everything else again btw.

**Comment by bohtho** [comment]
Actually.... I get the trouble with "Credential type 'sectoralarm-site' is not registered" and the alert window saying "Flows stopped due to missing node types. sectoralarm-site" whenever I have the Projects feature activated in Node-RED 0.18.4. I know the credentials file is handled differently then and also encrypted because of the Git repo and possible remotes (although not specified as such in settings.js). Can you verify ?

**Comment by perbrage** [comment]
I am using Node 0.17.5. It might be that my nodes are incompatible with that version, or maybe there is a bug in 0.18.4? I would need to upgrade my node-red environment to check that out.

Even so, I can't really produce the problem you are describing. Could you try to remove the node from all flows. Uninstall the NPM and then go through the files in your .node-red, and remove everything that has any sectoralarm name on it. If possible, remove the credentials file as well. Then reset NodeRed.

After that, please describe step-by-step what you are doing and I will try to reproduce the issue.

**Comment by bohtho** [comment]
Tried to remove cred file and npm uninstall but same problem as I installed sectoralarm in the node-red gui and configured the node. Now it has been three separate and clean node-red installations (all latest though) on different Debian distributions on two different machines. When I don’t activate the Projects/Git feature in node-red it seems fine, but as soon as it is activated I get the problem with credentials missing etc.

**Comment by bohtho** [comment]
I’m not saying for sure i’m not working in a strange and non-standard environment which messes up node-red, but it just seems to me there’s an incompatibility with either 0.18.4 or Projects.

**Comment by perbrage** [comment]
Ok! Thank you for the input, I will look into it

**Comment by perbrage** [comment]
Good and bad news, but

I upgraded Node-Red to 0.18.4. I don't know why the upgrade scripts does not upgrade to 0.18.5 but maybe it is too new for the raspberry pi scripts.

I turned on the setting for projects, and I created a project and my flows got converted and included into that project.

I went into projects settings, dependencies and clicked install on the dependent nodes (including this one). Deployed and restarted node red, and the node still works fine for me.

So it seems to work properly up to version 0.18.4 at least.

There is something that I want you to try, beyond making sure you installed the dependencies I described above.

1) Find the package.json (~/.node-red/node_modules/node-red-contrib-sectoralarm/package.json)
2) Edit the package.json, and in the bottom you can find a mapping:

Change:
"site": "lib/sectoralarm-site.js"

into:
"sectoralarm-site": "lib/sectoralarm-site.js"

3) Don't forget to restart node-red using something like

"sudo systemctl restart nodered"

That name discrepancy is the only one I can find in my code so far, and I would like to know if that change resolves the issue for you or not. If so, I can update that and push out a new version. Note, for me the node works no matter what I map it to site or sectoralarm-site.

**Comment by bohtho** [comment]
Good and bad news here too :)

Good for you though. I tried your package.json change, because indeed the references to both "sector alarm-site" and "site" could look like a discrepancy. But in my previous set-up that only displayed another identical config node besides the earlier one, but now I couldn't disable either of them. *Anyway*, I set up an entirely new DietPi SD card for my RPi3 without installing anything before node-red, *and* crucially, *not* using the rpi3 node-red-installation script which defaults to run as root, which I have accepted at the prompt. Now, on the other hand, the Sector Alarm node works beautifully and as expected. So I didn't even bother trying the package.js change again. It's fine.

So, sorry, but bottom-line is that there seemed to be an chown or chmod error I didn't notice I dragged with me through all installs, which happened to *only* affect my Sector Alarm node *after* I updated node-RED to Projects (which creates new dirs and moves files). Glad to have found that out at least, and that the sector node is nowhere to fault.
